### PR TITLE
Docker enhancements

### PR DESCRIPTION
**Pull request by FabienM**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Documentation | no
| Translation   | no
| Fixed tickets | None
| License       | MIT

Hi there,

As i was hacking around i ran into several problems when i wanted to get up a consistent docker stack, so i made some quick enhancements.

Hope this helps.

Regards,

###  Simplify docker environment variable management

* Leverage sf3.2 `%env()%` syntax
* Provide default values for non filled env variables
* Add redis and rabbitmq links to php container

### Enhance docker development enviromnent
    
* Switch to `app_dev.php` default front controller
* Allow docker default host IP to `app_dev.php`
* Add opcache and xdebug in php image

### CODE REVIEW

File: docs/en/developer/docker.rst [hunk 1]
```diff
@@ -30,15 +30,13 @@ In order to keep running Symfony commands on your host (such as
 ``wallabag:install``), you also should:
 
 - source the proper env files on your command line, so variables
-  like ``SYMFONY__ENV__DATABASE_HOST`` will exist.
+  like ``WALABAG_DATABASE_HOST`` will exist.
```
Comment (tcitworld) [inline docs/en/developer/docker.rst]: wallabag

File: web/app_dev.php [hunk 2]
```diff
@@ -12,7 +12,7 @@
 // Feel free to remove this, extend it, or make something more sophisticated.
 if (isset($_SERVER['HTTP_CLIENT_IP'])
     || isset($_SERVER['HTTP_X_FORWARDED_FOR'])
-    || !(in_array(@$_SERVER['REMOTE_ADDR'], ['127.0.0.1', 'fe80::1', '::1']) || php_sapi_name() === 'cli-server')
+    || !(in_array(@$_SERVER['REMOTE_ADDR'], ['127.0.0.1', '172.18.0.1', 'fe80::1', '::1']) || php_sapi_name() === 'cli-server')
```
Comment (tcitworld) [inline web/app_dev.php]: Not sure you should hardcode this

File: docker/nginx/nginx.conf [hunk 3]
```diff
@@ -41,7 +41,7 @@ http {
 
         location / {
             # try to serve file directly, fallback to app.php
-            try_files $uri /app.php$is_args$args;
+            try_files $uri /app_dev.php$is_args$args;
```
Comment (tcitworld) [inline docker/nginx/nginx.conf]: So `app_dev` is used as default ?

Comment (nicosomb) [inline docker/nginx/nginx.conf]: These Docker files are here to have a dev environment with Docker, not for production use. So why not setting `app_dev` here? 
For production environment, you need to use the Docker image.

File: web/app_dev.php (hunk 2 above)
Comment (FabienM) [inline web/app_dev.php]: Same "hardcoding" policy that for `127.0.0.1`, it's a default value for docker network brigde on a private subnet.
The purpose is to allow docker user to directly access `app_dev.php` without having to modify it.

Could also be replaced with documentation.

File: docker/nginx/nginx.conf (hunk 3 above)
Comment (tcitworld) [inline docker/nginx/nginx.conf]: Was lost in repositories...

Comment (FabienM) [inline docker/nginx/nginx.conf]: Yup, it's my use case. Maybe you have another one for this docker-compose that should be considered?

Same logic that enabling xdebug on php image. Wouldn't have made sense on a production image :-)

**Comment by FabienM** [comment]
Probably, indeed, to make better use of `SYMFONY__ENV__DATABASE_PORT`~ish variables.
I'll take a look.

File: docker/nginx/nginx.conf (hunk 3 above)
Comment (nicosomb) [inline docker/nginx/nginx.conf]: You can continue on these files 👍

**Comment by nicosomb** [comment]
Ready to be merged?

**Comment by nicosomb** [comment]
ping @FabienM

**Comment by FabienM** [comment]
Hmm, I suppose... Should I check something more ?

**Comment by FabienM** [comment]
Rebased on master, just in case.

**Comment by nicosomb** [comment]
I will add documentation.

**Comment by FabienM** [comment]
Hi,

What seems wrong for you?

```bash
$ rm app/config/parameters.yml
rm : supprimer fichier 'app/config/parameters.yml' ? y
$ c install --ignore-platform-reqs
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Nothing to install or update
Generating autoload files
> Incenteev\ParameterHandler\ScriptHandler::buildParameters
Creating the "app/config/parameters.yml" file
<...incenteev chitchat...>
$ cat app/config/parameters.yml
# This file is auto-generated during the composer install
parameters:
    database_driver: '%env(WALLABAG_DATABASE_DRIVER)%'
    database_host: '%env(WALLABAG_DATABASE_HOST)%'
    database_port: '%env(WALLABAG_DATABASE_PORT)%'
    database_name: '%env(WALLABAG_DATABASE_NAME)%'
    database_user: '%env(WALLABAG_DATABASE_USER)%'
    database_password: '%env(WALLABAG_DATABASE_PASSWORD)%'
    database_path: '%env(WALLABAG_DATABASE_PATH)%'
    database_table_prefix: wallabag_
...
```

Everything looks good to me, the build ends properly and instance is reachable through docker.

**Comment by nicosomb** [comment]
> reachable through docker

Through docker, maybe 👍 
but not through a classic web server. Try to launch the built in server (`server:run`).

**Comment by FabienM** [comment]
Working as well :-/

```
$ php bin/console server:run 
                                                             
 [OK] Server running on http://127.0.0.1:8000                                                                           
                                                                      
 // Quit the server with CONTROL-C.                                                                                     

PHP 7.0.15-0ubuntu0.16.04.4 Development Server started at Mon Mar 20 11:44:26 2017
Listening on http://127.0.0.1:8000
Document root is /home/fameu/Workspaces/FabienM/wallabag/web
Press Ctrl-C to quit.
[Mon Mar 20 11:44:44 2017] 127.0.0.1:60016 [302]: /
[Mon Mar 20 11:44:44 2017] 127.0.0.1:60018 [200]: /login
```

That's the point of `parameters_default.yml`, valorize parameters that are not by the environment.
Is this file still imported inyour `config.yml`?

Are you able to run `bin/console debug:container --parameters | grep 'env('`?

**Comment by nicosomb** [comment]
Maybe I was drunk. 
I think I forgot the `wallabag:install` step. 

So it's OK 👍  But as I told you privately, the parameters step is a bit confusing: 

capture d ecran 2017-03-23 a 14 15 07

Today, the default value (pdo_sqlite for example) is displayed.

**Comment by FabienM** [comment]
Were you suggesting that I didn't test my code ? :trollface: 

What would fit you ? I can propose to drop those parameters from `parameters.yml.dist` so the environment variables are the default way to configure Wallabag and the Incenteev wizard is no more confusing (and lighter for newbees).
It's also consistent with Symfony current orientation and the future Dotenv component.

I have no other solution to make this work out of the box with docker AND embedded server while preserving default values in `parameters.yml.dist`.

If this is too confusing, the only solution i see is to version a `parameters.yml.env` or something and tell the developer to manually import this file if needed.

**Comment by nicosomb** [comment]
> What would fit you ? I can propose to drop those parameters from parameters.yml.dist so the environment variables are the default way to configure Wallabag and the Incenteev wizard is no more confusing

👍 It looks for me. 

@wallabag/core any opinion?

**Comment by nicosomb** [comment]
@FabienM Do you think that you'll have time to work on parameters step?

**Comment by FabienM** [comment]
i'll try to give a look soon

**Comment by j0k3r** [comment]
👋 ?

File: docker-compose.yml [hunk 4]
```diff
@@ -22,9 +22,11 @@ services:
             - "9000:9000"
         volumes:
             - .:/var/www/html
-        #links:
+        links:
             # - "postgres:rdbms"
             # - "mariadb:rdbms"
+            - rabbitmq
+            - reddis
```
Comment (aaa2000) [inline docker-compose.yml]: `reddis` need to be change to `redis`

**Comment by nicosomb** [comment]
@FabienM 👋

**Comment by FabienM** [comment]
I really would like to help but unfortunately I do not have a ready to use PHP/docker environment and not much time to create it (can't use my professional laptop anymore).

A quick win would be to fix the `reddis` thing, remove doc (and create a PR on the proper repo?) and merge this one.

What do you think?

**Comment by nicosomb** [comment]
I'm OK for that. 
And I can create the new PR on the documentation repo if you want.

**Comment by j0k3r** [comment]
This PR is fairly old and there hasn't been much activity on it.
Closing, but please re-open if you have time to work on it.

**Pull request closed without merging**